docs/ref/models/expressions.txt

```diff
@@ -640,3 +640,28 @@ Let's see how it works::
     Apple: AAPL
     Yahoo: Internet Company
     Django Software Foundation: No Tagline
+
+Adding support for third party backends
+---------------------------------------
+If you are using a backend that uses a different syntax for a certain function,
+you can easily add support for it by monkey patching a new method onto the
+relevant expression class.
+
+Let's say we are using the `Firebird` backend. If `Firebird` would use a
+different syntax for the ``SUM`` function, we could monkey patch a new method
+called ``as_firebird`` onto the ``Sum`` class. We use ``as_firebird`` because
+``connection.vendor`` returns ``firebird`` when using the `Firebird` backend.
+
+In our new method, we can override the default value of ``template`` or
+``function``, or even do more complicated things.
+
+In our example, we could end up with something like this::
+
+    from django.db.models import Sum
+
+    def firebird_sum(self, compiler, connection):
+        template = 'SUM(%(expressions)s)'
```

Comment:
`as_sql` does not accept a `template` override, or any `*args` or `**kwargs` for that matter. For the moment, an `as_vendor` override has to either set self.template (I'll come back to this), or clone self, and change the `clone.template`.

@carljm is right regarding shared expressions on different backends. I realised this when writing my talk that it's a dangerous place for modifying self.

To solve both problems, I think we should adjust `as_sql` to accept `**kwargs` that can be mixed with `self.extra`. That's not a job for this patch though, so I'll open a separate ticket.

This patch should recommend setting `self.template` for now (or clone.template probably?), and whoever implements the as_sql kwargs patch can also update these recommendations to use kwargs.